### PR TITLE
perf: Speed up SDK init

### PR DESCRIPTION
**Pull request by jairbubbles**
In some scenarios the time spent in initializing the SDK can be too important. What is tricky is that the init of the SDK is one of the first thing the app do and we're JITing almost everything. Creating a `HttpClient`, a `Regex` is costly...

In this pull request I'm experimenting with lazy instanciation so all that JIT time for these types will be spent afterwards.

I'm disabling almost everything to be in the best scenario possible:
```csharp
await Task.Run(() =>
{
    SentrySdk.Init(o =>
    {
        o.Dsn = "<sentry_dsn>";
        o.Release = "1.0.0";
        o.DisableAppDomainProcessExitFlush();
        o.DisableAppDomainUnhandledExceptionCapture();
        o.DisableUnobservedTaskExceptionCapture();
        o.DisableDuplicateEventDetection();
        o.DisableDiagnosticSourceIntegration();
        o.DisableWinUiUnhandledExceptionIntegration(); // this one is new!
        o.AutoSessionTracking = false;
        o.IsGlobalModeEnabled = true;
        o.DetectStartupTime = StartupTimeDetectionMode.Fast;
        o.TracesSampleRate = 0;
    });
});
```

Here is the current status:

image

Compared to main branch:

image

Let me know what you think!

### CODE REVIEW

File: src/Sentry/SentryOptions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:71]: I think this is called often so we could cache the value here on first access:

```suggestion
        get => _isGlobalModeEnabled ??= Runtime.Current.IsBrowserWasm();
```

File: src/Sentry/SentryOptionsExtensions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptionsExtensions.cs:354]: no functional change here, just style?

File: src/Sentry/Internal/Http/LazyHttpTransport.cs
Comment (bruno-garcia) [inline src/Sentry/Internal/Http/LazyHttpTransport.cs:18]: How about moving this all under a `Lazy<>` and access it via `_lazyTransport.Value` ?

File: src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs
Comment (bruno-garcia) [inline src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs:18]: Now with the type, we prob should go with a dictionary Type:instance
We're doing allocations in each of the LINQ calls, and iterating the lists

**Review by bruno-garcia — COMMENTED** [review]
Nice! Thanks for this PR, the idea makes sense. I realize it's a draft but left some notes already.

Could you please add some benchmarks in https://github.com/getsentry/sentry-dotnet/tree/main/benchmarks ?

File: src/Sentry/SentryOptionsExtensions.cs
Comment (jairbubbles) [reply src/Sentry/SentryOptionsExtensions.cs:354]: It's because of successive refactor. I'll fix it.

File: src/Sentry/Internal/Http/LazyHttpTransport.cs
Comment (jairbubbles) [reply src/Sentry/Internal/Http/LazyHttpTransport.cs:18]: Sure, there's not enough `Lazy` in my PR 😅

File: src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs
Comment (jairbubbles) [reply src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs:18]: That was my previous modification but I've had issues in unit test where we use substitutes as they were sharing the same "RuntimeType"  type.

Also, it was not clear to me if we can or not have several times the same types.

At the end, I use `List<Tuple>` as I felt like it was more close to previous code.

**Comment by jairbubbles** [comment]
> Could you please add some benchmarks in https://github.com/getsentry/sentry-dotnet/tree/main/benchmarks ?

I struggled a bit to add the benchmarks as I needed a specific "ColdStart" configuration but here we go:

|            Method |     Mean |    Error |   StdDev |      Min |      Max |   Median | Methods JITted | Methods Tiered | JIT allocated memory | Allocated |
|------------------ |---------:|---------:|---------:|---------:|---------:|---------:|---------------:|---------------:|---------------------:|----------:|
| 'Init SDK (Full)' | 68.88 ms | 63.86 ms | 16.59 ms | 58.46 ms | 98.09 ms | 63.95 ms |          1,160 |            110 |               236 KB |  44.35 KB |
| 'Init SDK (Slim)' | 44.72 ms | 52.51 ms | 13.64 ms | 35.17 ms | 66.58 ms | 36.53 ms |            900 |              1 |               165 KB |  28.34 KB |

**Comment by jairbubbles** [comment]
Some notes:
- `Dns.Parse` is still taking a bit of time (20ms), maybe we could do a simple verification in the constructor and lazy init it?
- Some features which I disabled in my use case were taking a lot of time : in particular AutoSession is writing / reading a file for the persistant id.

**Comment by vaind** [comment]
> Dns.Parse is still taking a bit of time (20ms),

that's not a bit of time, that's a lot of time for such a simple operation (seemingly). Could use a bit more investigation.

**Comment by jairbubbles** [comment]
> > Dns.Parse is still taking a bit of time (20ms),
> 
> that's not a bit of time, that's a lot of time for such a simple operation (seemingly). Could use a bit more investigation.

Yeah but it's due to JIT.

**Comment by bruno-garcia** [comment]
> > > Dns.Parse is still taking a bit of time (20ms),
> > 
> > 
> > that's not a bit of time, that's a lot of time for such a simple operation (seemingly). Could use a bit more investigation.
> 
> Yeah but it's due to JIT.

We looked at it with @bitsandfoxes when profiling this and IIRC it was mainly the `Uri` class, most time spent.
If that's the case we could prob stop using it and take a simpler approach.
We parse it in other SDKs, e.g in `C`: https://github.com/getsentry/sentry-native/blob/406d41881111d15b1144b443d51760c1415c8029/src/sentry_utils.c#L216-L270

File: src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs
Comment (bruno-garcia) [reply src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs:18]: Fair, Tuple is the way to go unless we'd want to break that behavior.
Still dropping the LINQ usage is potential area of improvement if we're looking for more.

Comment (jairbubbles) [reply src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs:18]: Still, overall it feels a bit complex. How would you feel if those default integrations would be stored in an enum flag? During the config you would just need to remove the flags in the `DisableXXX` methods. It requires a bit more work though.

Comment (jairbubbles) [reply src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs:18]:
```csharp
[Flags]
internal enum DefaultIntegrations
{
    AutoSessionTrackingIntegration = 1 << 0,
    AppDomainUnhandledExceptionIntegration = 1 << 1,
    AppDomainProcessExitIntegration = 1 << 2,
    UnobservedTaskExceptionIntegration = 1 << 3,
    NetFxInstallationsIntegration = 1 << 4,
    SentryDiagnosticListenerIntegration = 1 << 5,
    WinUiUnhandledExceptionIntegration = 1 << 6,
}
```

Comment (jairbubbles) [reply src/Sentry.DiagnosticSource/SentryOptionsDiagnosticExtensions.cs]: I gave it a shot in my last commit.

File: src/Sentry/SentryOptions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:1237]: All default integrations are internal, by finding its usages we could spot any removal of default integrations 

```suggestion
    // Note: Not removing default integrations
```

We have usage in Unity (which access internal API) but no default integration there: https://github.com/search?q=repo%3Agetsentry%2Fsentry-unity%20RemoveIntegration&type=code


A comment on the method stating it doesn't apply to default integrations would be good enough I imagine

File: src/Sentry/Internal/Hub.cs
Comment (bruno-garcia) [inline src/Sentry/Internal/Hub.cs:65]: Could we get around iterating on `Integrations` below without allocating the list?

**Review by bruno-garcia — COMMENTED** [review]
This is quite promising.Once it's out of draft we can test it against `feat/4.0.0` to make sure it works fine with AOT

**Comment by bruno-garcia** [comment]
> But it's not so convincing when running it on main to compare:

Would be more noticible if we had some "application code" right after Init. That now would get a chance to run before some gets sent into Sentry vs before where it would need everything to bootstrap

File: src/Sentry/Internal/Hub.cs
Comment (jairbubbles) [reply src/Sentry/Internal/Hub.cs:65]: My first draft didn't have the `ToList()` and some test was failing. One integration was removing another integration.

It's probably this one (which will work fine now with the enum flags):
```csharp
internal class SentryDiagnosticListenerIntegration : ISdkIntegration
{
    public void Register(IHub hub, SentryOptions options)
    {
        if (!options.IsPerformanceMonitoringEnabled)
        {
            options.Log(SentryLevel.Info, "DiagnosticSource Integration is disabled because tracing is disabled.");
            options.DisableDiagnosticSourceIntegration();
            return;
        }

        var subscriber = new SentryDiagnosticSubscriber(hub, options);
        DiagnosticListener.AllListeners.Subscribe(subscriber);
    }
}
```

Comment (jairbubbles) [reply src/Sentry/Internal/Hub.cs]: In fact no it won't be working, there is a duality with `SentryDiagnosticListenerIntegration` as it's internaly provided or provided by `Sentry.DiagnosticSource`. 

I feel like it's useless to remove itself from the options. I'm removing the `ToList`, I just adapted a test to check the log instead.

**Comment by jairbubbles** [comment]
With latest changes, here are the benchmarks results:

|            Method |     Mean |    Error |   StdDev |      Min |      Max |   Median | Methods JITted | Methods Tiered | JIT allocated memory | Allocated |
|------------------ |---------:|---------:|---------:|---------:|---------:|---------:|---------------:|---------------:|---------------------:|----------:|
| 'Init SDK (Full)' | 42.91 ms | 1.679 ms | 1.110 ms | 41.51 ms | 44.63 ms | 42.52 ms |          1,028 |              1 |               186 KB |  43.05 KB |
| 'Init SDK (Slim)' | 25.45 ms | 0.981 ms | 0.649 ms | 24.51 ms | 26.25 ms | 25.56 ms |            891 |              1 |               161 KB |  27.22 KB |

**Review by bruno-garcia — APPROVED** [review]
Great stuff! Thanks @jairbubbles 
Prob needs @bitsandfoxes @vaind or @jamescrosswell as some extra check before we can merge but lgtm

**Comment by bitsandfoxes** [comment]
Those are some serious gains! I'll follow up with a review/merge this tomorrow.

**Review by bitsandfoxes — APPROVED** [review]
Let me run this against `4.0.0` before merging to avoid any heartache.

**Comment by bruno-garcia** [comment]
Lets merge this and figure out AOT after

**Comment by bruno-garcia** [comment]
Thanks @jairbubbles

**Comment by bitsandfoxes** [comment]
Already tested. Works like a charm. No hickups there!

**Pull request merged**
